File: code_snippets/python/src/use_cases/event_processing/package_tracker.py
from typing import List

import restate
from pydantic import BaseModel
from restate import VirtualObject, ObjectContext
from restate.exceptions import TerminalError


class LocationUpdate(BaseModel):
    location: str
    timestamp: str


class PackageInfo(BaseModel):
    final_destination: str
    locations: List[LocationUpdate] = []


# <start_here>
# <mark_2>
package_tracker = VirtualObject("package-tracker")
# </mark_2>


# <mark_3>
# <mark_2>
@package_tracker.handler()
async def register_package(ctx: ObjectContext, package_info: PackageInfo):
    # </mark_2>
    # </mark_3>
    # <mark_1>
    ctx.set("package-info", package_info.model_dump())
    # </mark_1>


# <mark_3>
# <mark_2>
@package_tracker.handler()
async def update_location(ctx: ObjectContext, location_update: LocationUpdate):
    # </mark_2>
    # </mark_3>
    # <mark_1>
    package_info = await ctx.get("package-info")
    # </mark_1>
    if package_info is None:
        raise TerminalError(f"Package {ctx.key()} not found")

    locations = package_info.get("locations", [])
    locations.append(location_update.model_dump())
    package_info["locations"] = locations

    # <mark_1>
    ctx.set("package-info", package_info)
    # </mark_1>


# <mark_3>
# <mark_2>
@package_tracker.handler()
async def get_package_info(ctx: ObjectContext):
    # </mark_2>
    # </mark_3>
    # <mark_1>
    return await ctx.get("package-info")
    # </mark_1>


app = restate.app(services=[package_tracker])
# <end_here>

File: code_snippets/kotlin/src/main/kotlin/usecases/eventprocessing/PackageTracker.kt
package usecases.eventprocessing

import dev.restate.sdk.annotation.Handler
import dev.restate.sdk.annotation.Shared
import dev.restate.sdk.annotation.VirtualObject
import dev.restate.sdk.common.TerminalException
import dev.restate.sdk.http.vertx.RestateHttpServer
import dev.restate.sdk.kotlin.*
import dev.restate.sdk.kotlin.endpoint.endpoint
import kotlinx.serialization.Serializable

// <start_here>
// <mark_2>
@VirtualObject
class PackageTracker {
  // </mark_2>

  companion object {
    private val PACKAGE_INFO = stateKey<PackageInfo>("package-info")
  }

  // <mark_3>
  // <mark_2>
  @Handler
  suspend fun registerPackage(ctx: ObjectContext, packageInfo: PackageInfo) {
    // </mark_2>
    // </mark_3>
    // <mark_1>
    ctx.set(PACKAGE_INFO, packageInfo)
    // </mark_1>
  }

  // <mark_3>
  // <mark_2>
  @Handler
  suspend fun updateLocation(ctx: ObjectContext, locationUpdate: LocationUpdate) {
    // </mark_2>
    // </mark_3>
    // <mark_1>
    val packageInfo =
        ctx.get(PACKAGE_INFO)
            // </mark_1>
            ?: throw TerminalException("Package not found")

    packageInfo.locations.add(locationUpdate)
    // <mark_1>
    ctx.set(PACKAGE_INFO, packageInfo)
    // </mark_1>
  }

  // <mark_3>
  // <mark_2>
  @Shared
  suspend fun getPackageInfo(ctx: SharedObjectContext): PackageInfo {
    // </mark_2>
    // </mark_3>
    // <mark_1>
    return ctx.get(PACKAGE_INFO)
        // </mark_1>
        ?: throw TerminalException("Package not found")
  }
}

fun main() {
  RestateHttpServer.listen(endpoint { bind(PackageTracker()) })
}
// <end_here>

@Serializable
data class PackageInfo(
    val finalDestination: String,
    val locations: MutableList<LocationUpdate> = mutableListOf()
)

@Serializable data class LocationUpdate(val timestamp: String, val location: String)
